File: sample-code/graphs/search/python/graph_search.py
def main():
    graph = read_graph_from_file("graph01.txt")
    print bfs(graph, graph.vertices[0])
    print dfs(graph, graph.vertices[0])


def read_graph_from_file(filename):
    with open(filename, "r") as input_file:
        # The first two lines define the vertex count and the edge count.
        vertex_count = int(input_file.readline())
        edge_count = int(input_file.readline())

        vertices = [Vertex(index) for index in range(0, vertex_count)]
        edges = []

        # The next 'edge_count' lines describe the edges: "start_vertex end_vertex".
        for i in range(0, edge_count):
            start_vertex, end_vertex = [int(index) for index in input_file.readline().split()]

            # Adding the edge to the list of outbound edges for the start vertex.
            edge = Edge(vertices[start_vertex], vertices[end_vertex])
            vertices[start_vertex].outbound_edges.append(edge)

            # For non-directed graphs, an outbound edge is also an inbound one (0 -> 1 == 1 -> 0).
            # Therefore, we reverse the edge and add it to the other vertex.
            reverse_edge = Edge(vertices[end_vertex], vertices[start_vertex])
            vertices[end_vertex].outbound_edges.append(reverse_edge)

            edges.append(edge)
            edges.append(reverse_edge)

        return Graph(vertices, edges)


def bfs(graph, start_vertex):
    result = []

    # Initially, the queue contains only the start vertex
    # and all vertices are assumed to be not visited yet.
    queue = [start_vertex]
    visited = [False for _ in range(0, len(graph.vertices))]

    while len(queue) > 0:
        # Remove a vertex from the queue.
        current_vertex = queue.pop(0)

        # If we've already been here, ignoring this vertex completely.
        # This condition can happen when, for example, this vertex was a neighbor of
        # two other vertices and they both added it to the queue before it was visited.
        if visited[current_vertex.label]:
            continue

        # Otherwise, marking it as visited so that we won't analyze it anymore.
        visited[current_vertex.label] = True

        # Getting all adjacent vertices which haven't been visited yet.
        # It's only a matter of traversing the outbound_edges list and getting end_vertex for each.
        neighbors = [edge.end_vertex
                     for edge in current_vertex.outbound_edges
                     if not visited[edge.end_vertex.label]]

        # If we need to enforce a particular ordering on the neighbors we visit,
        # e.g., visit them in the order of increasing labels (1, 4, 6; not 4, 6, 1),
        # this would be the place to do the sorting.

        # Adding these neighbors to the queue, all at once.
        result.append(current_vertex.label)
        queue.extend(neighbors)

    return result


# See the comments for BFS above.
# Note that the ONLY difference between BFS and DFS is using a queue vs. using a stack!
# In this Python implementation, it's only a matter of calling:
#    list.pop(0) (remove the leftmost item -> queue behavior, hence BFS)
#    list.pop()  (remove the rightmost item -> stack behavior, hence DFS)
def dfs(graph, start_vertex):
    result = []

    stack = [start_vertex]
    visited = [False for _ in range(0, len(graph.vertices))]

    while len(stack) > 0:
        current_vertex = stack.pop()
        if visited[current_vertex.label]:
            continue

        visited[current_vertex.label] = True
        neighbors = [edge.end_vertex
                     for edge in current_vertex.outbound_edges
                     if not visited[edge.end_vertex.label]]

        result.append(current_vertex.label)
        stack.extend(neighbors)

    return result


class Vertex:
    def __init__(self, label):
        self.label = label
        self.outbound_edges = []

    def __str__(self):
        return "Label: %d    Edges: %s" % (self.label, ', '.join([str(edge) for edge in self.outbound_edges]))


class Edge:
    def __init__(self, start_vertex, end_vertex):
        self.start_vertex = start_vertex
        self.end_vertex = end_vertex

    def __str__(self):
        return "%d -> %d" % (self.start_vertex.label, self.end_vertex.label)


class Graph:
    def __init__(self, vertices, edges):
        self.vertices = vertices
        self.edges = edges


if __name__ == "__main__":
    main()

File: sample-code/graphs/search/java/src/GraphSearch.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.*;
import java.util.stream.*;

public class GraphSearch {

    public static void main(String[] args) throws FileNotFoundException {
        Graph graph = readGraphFromFile("graph01.txt");
        List<Vertex> bfsVertices = bfs(graph, graph.getVertices().get(0));
        List<Vertex> dfsVertices = dfs(graph, graph.getVertices().get(0));

        System.out.println("BFS traversal order:");
        System.out.println(bfsVertices.stream()
                .map(v -> String.valueOf(v.getLabel()))
                .collect(Collectors.joining(", ")));
        System.out.println();

        System.out.println("DFS traversal order:");
        System.out.println(dfsVertices.stream()
                .map(v -> String.valueOf(v.getLabel()))
                .collect(Collectors.joining(", ")));
    }

    private static Graph readGraphFromFile(String fileName) throws FileNotFoundException {
        File inputFile = new File(fileName);

        try (Scanner inputFileScanner = new Scanner(inputFile)) {
            // The first two lines define the vertex count and the edge count.
            int vertexCount = inputFileScanner.nextInt();
            int edgeCount = inputFileScanner.nextInt();

            List<Vertex> vertices = new ArrayList<>(vertexCount);
            List<Edge> edges = new ArrayList<>(edgeCount);

            for (int v = 0; v < vertexCount; v++) {
                vertices.add(v, new Vertex(v));
            }

            // The next 'edgeCount' lines describe the edges: "startVertex endVertex".
            for (int e = 0; e < edgeCount; e++) {
                int startVertexLabel = inputFileScanner.nextInt();
                int endVertexLabel = inputFileScanner.nextInt();

                Edge edge = new Edge(vertices.get(startVertexLabel), vertices.get(endVertexLabel));
                edges.add(edge);

                // Adding the edge to the list of outbound edges for the start vertex.
                vertices.get(startVertexLabel).getOutboundEdges().add(edge);

                // For non-directed graphs, an outbound edge is also an inbound one (0 -> 1 == 1 -> 0).
                // Therefore, we reverse the edge and add it to the other vertex.
                Edge reverseEdge = new Edge(vertices.get(endVertexLabel), vertices.get(startVertexLabel));
                edges.add(reverseEdge);
                vertices.get(endVertexLabel).getOutboundEdges().add(reverseEdge);
            }

            return new Graph(vertices, edges);
        }
    }

    private static List<Vertex> bfs(Graph graph, Vertex startVertex) {
        List<Vertex> traversalOrder = new ArrayList<>();

        // Initially, the queue contains only the start vertex
        // and all vertices are assumed to be not visited yet.
        Queue<Vertex> queue = new ArrayDeque<>();
        queue.add(startVertex);

        boolean[] isVisited = new boolean[graph.getVertices().size()];

        while (queue.size() > 0) {
            // Remove a vertex from the queue.
            Vertex currentVertex = queue.remove();

            // If we've already been here, ignoring this vertex completely.
            // This condition can happen when, for example, this vertex was a neighbor of
            // two other vertices and they both added it to the queue before it was visited.
            if (isVisited[currentVertex.getLabel()]) {
                continue;
            }

            // Otherwise, marking it as visited so that we won't analyze it anymore.
            isVisited[currentVertex.getLabel()] = true;

            // Getting all adjacent vertices which haven't been visited yet.
            // It's only a matter of traversing the outboundEdges list and getting endVertex for each.
            List<Vertex> unvisitedNeighbors = new ArrayList<>();
            for (Edge outboundEdge: currentVertex.getOutboundEdges()) {
                if (!isVisited[outboundEdge.getEndVertex().getLabel()]) {
                    unvisitedNeighbors.add(outboundEdge.getEndVertex());
                }
            }

            // If we need to enforce a particular ordering on the neighbors we visit,
            // e.g., visit them in the order of increasing labels (1, 4, 6; not 4, 6, 1),
            // this would be the place to do the sorting.

            // Adding the unvisited neighbors to the queue, all at once.
            queue.addAll(unvisitedNeighbors);

            // Adding the current vertex to the overall result.
            traversalOrder.add(currentVertex);
        }

        return traversalOrder;
    }

    private static List<Vertex> dfs(Graph graph, Vertex startVertex) {
        // See the comments for BFS above.
        // Note that the ONLY difference between BFS and DFS is using a queue vs. using a stack!
        List<Vertex> traversalOrder = new ArrayList<>();

        Stack<Vertex> stack = new Stack<>();
        stack.push(startVertex);
        boolean[] isVisited = new boolean[graph.getVertices().size()];

        while (stack.size() > 0) {
            Vertex currentVertex = stack.pop();

            if (isVisited[currentVertex.getLabel()]) {
                continue;
            }
            isVisited[currentVertex.getLabel()] = true;

            List<Vertex> unvisitedNeighbors = new ArrayList<>();
            for (Edge outboundEdge: currentVertex.getOutboundEdges()) {
                if (!isVisited[outboundEdge.getEndVertex().getLabel()]) {
                    unvisitedNeighbors.add(outboundEdge.getEndVertex());
                }
            }

            stack.addAll(unvisitedNeighbors);
            traversalOrder.add(currentVertex);
        }

        return traversalOrder;
    }
}
